src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -202,33 +213,92 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
                 throw new ArgumentOutOfRangeException(nameof(serverAddresses));
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            var sockets = new List<Socket>(serverAddresses.Length);
+            var connectTasks = new List<Task>(serverAddresses.Length);
+            var tcs = new TaskCompletionSource<Socket>();
+            var lastError = new StrongBox<Exception>();
+            var pendingCompleteCount = new StrongBox<int>(serverAddresses.Length);
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                // Start all connection tasks now, to prevent possible race conditions with
+                // calling ConnectAsync on disposed sockets.
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    var socket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
+                    Task connectTask = socket.ConnectAsync(address, port);
+
+                    sockets.Add(socket);
```

Comment:
I was hitting issues before where a successful connection would dispose the other sockets, leading to an ObjectDisposed exception when trying to call socket.ConnectAsync in later tasks. While that exception could be caught, I figured it would be better to avoid generating exceptions if possible